### ISSUE TITLE
CDNReourseHandler not get resource with omnifaces 2.0

### ISSUE THREAD

**Issue by CatWingHub** [omnifaces/omnifaces#85]
Hi!

with env:
TomEE1.7.1(apache-tomee-webprofile-1.7.1)
myfaces 2.2.4/mojjara 2.2.9

pom.xml:
```xml
	<dependencies>
		<dependency>
			<groupId>org.glassfish</groupId>
			<artifactId>javax.faces</artifactId>
			<version>2.2.9</version>
		</dependency>
		<dependency>
			<groupId>org.omnifaces</groupId>
			<artifactId>omnifaces</artifactId>
			<version>2.0</version>
		</dependency>
	</dependencies>
```
web.xml:
```xml
    <context-param>
        <description>
            Tell the runtime where we are in the project development
            lifecycle.  Valid values are: 
            Development, UnitTest, SystemTest, or Production.
            The runtime will display helpful hints to correct common mistakes
            when the value is Development.
        </description>
        <param-name>javax.faces.PROJECT_STAGE</param-name>
        <param-value>Production</param-value>
    </context-param>
    <context-param>
    	<param-name>org.omnifaces.CDN_RESOURCE_HANDLER_URLS</param-name>
     	<param-value>
         	testlib:js/jquery1.js=http://ajax.aspnetcdn.com/ajax/jQuery/jquery-2.1.1.js
     	</param-value>
 	</context-param>
```
faces-config.xml:
```xml
<faces-config version="2.0"
    xmlns="http://java.sun.com/xml/ns/javaee"
    xmlns:xi="http://www.w3.org/2001/XInclude"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-facesconfig_2_0.xsd">
	<application>
    	<resource-handler>org.omnifaces.resourcehandler.CDNResourceHandler</resource-handler>
    </application>
	
</faces-config>
```

page fragment:
```xml
<h:head>
  <h:outputScript name="js/jquery1.js" library="testlib"></h:outputScript>
</h:head>
```
when migrate jsf version from 2.1.7 to 2.2.x(myfaces 2.2.4 and mojjara 2.2.9 tested) with update omnifaces from 1.8.1 to 2.0, found cdn resource string configed in web.xml not rendered with console output:
```console
Resource referenced by resourceName js/jquery1.js and libraryName testlib not found in call to ResourceHandler.createResource. It will be silenty ignored.
```
with myfaces 2.1.7 and omnifaces 1.8.1 can work.
after update myfaces to 2.2.4 with omnifaces 2.0, the issue appeared with console output above.
Then have tried mojjara 2.2.9, and also try to create a new skeleton project for testing it as config above, the result still keep same.

Debug with code CDNResourceHandler:
```java
	@Override
	public Resource createResource(String resourceName, String libraryName, String contentType) {
		Resource resource = super.createResource(resourceName, libraryName, contentType);

		if (resource == null || (disabledParam != null && Boolean.valueOf(String.valueOf(evaluateExpressionGet(disabledParam))))) {
			return resource;
		}

		String requestPath = null;

		if (cdnResources != null) {
			requestPath = cdnResources.get(new ResourceIdentifier(libraryName, resourceName));

			if (requestPath == null) {
				requestPath = cdnResources.get(new ResourceIdentifier(libraryName, "*"));

				if (requestPath != null) {
					requestPath = requestPath.substring(0, requestPath.length() - 1) + resourceName;
				}
			}
		}

		if (requestPath == null) {
			return resource;
		}

		final String evaluatedRequestPath = evaluateExpressionGet(requestPath);

		return new DefaultResource(resource) {
			@Override
			public String getRequestPath() {
				return evaluatedRequestPath;
			}
		};
	}
```
and tried several config pattern as in http://showcase.omnifaces.org/resourcehandlers/CDNResourceHandler,
found 
```java
Resource resource = super.createResource(resourceName, libraryName, contentType);

		if (resource == null || (disabledParam != null && Boolean.valueOf(String.valueOf(evaluateExpressionGet(disabledParam))))) {
			return resource;
		}
```
resource will always get null, then return without reaching the login of cdn resource render.

Could you please help to have a check for it or it's just because I have some config not match with omnifaces update? Thanks a lot!

**Comment by BalusC** [comment]
Can you please try the current OmniFaces 2.1 SNAPSHOT? https://oss.sonatype.org/content/repositories/snapshots/org/omnifaces/omnifaces/2.1-SNAPSHOT/

There was a change in the way the resource handler delegates the call to the wrapped one, which turned out to be somewhat problematic when used in combination with multiple component libraries such as PrimeFaces + PrimeFaces-Extensions. This has been improved for OmniFaces 2.1.

**Comment by BalusC** [comment]
Any progress?

**Comment by djmj** [comment]
I tried migrating from 1.8.1 to todays snapshot of 2.1 

I removed org.omnifaces.CDN_RESOURCE_HANDLER_ALWAYS_ENABLED but resources could not be resolved.

Same problem as:

http://stackoverflow.com/questions/28485512/cdnresourcehandler-stopped-working-in-omnifaces-2-0

**Comment by BalusC** [comment]
Which resources exactly are you CDN'ing? Do those resources exist in the webap? Is the JSF project stage set to production?

**Comment by djmj** [comment]
No they did not exist in the webapp (or most of them) and it was set to development.

**Comment by BalusC** [comment]
The `CDNResourceHandler` is intented to move auto-included JSF resources to a CDN.

If you don't have a JSF resource, just use plain HTML `<link>` or `<script>`.

**Comment by djmj** [comment]
The advantage of the `old` `CDNResourceHandler` for the developer was not to worry where the resource comes from and use JSF library syntax. Now we must worry and pollute the markup with url's and paths with html tags.

Also when changes of resource location occur (from within webapp to outside or vice versa) we need  to update markup.

Is there a possibility to provide backwards compatibility?

**Comment by CatWingHub** [comment]
Hi, sorry for the late reply..
Have tried 2.1 SNAPSHOT but found it still not working as:
```html
<script type="text/javascript" src="RES_NOT_FOUND"></script>
<script type="text/javascript" src="RES_NOT_FOUND"></script>
<link type="text/css" rel="stylesheet" href="RES_NOT_FOUND">
```
have tried the CDN config with below and the other config is same as this post above:

```xml
<context-param>
    <param-name>org.omnifaces.CDN_RESOURCE_HANDLER_URLS</param-name>
    <param-value>
        bootstrap.min.css=http://cdn.bootcss.com/bootstrap/3.3.0/css/bootstrap.min.css,
        bootstrap.min.js=http://apps.bdimg.com/libs/bootstrap/3.3.0/js/bootstrap.min.js,
        jquery.js=http://ajax.aspnetcdn.com/ajax/jQuery/jquery-2.1.1.js
    </param-value>
</context-param>
```
```html
<head>
	<h:outputStylesheet name="bootstrap.min.css" />
	<h:outputScript name="jquery.js" />
	<h:outputScript name="bootstrap.min.js" />
</head>
```
and
```xml
<context-param>
    <param-name>org.omnifaces.CDN_RESOURCE_HANDLER_URLS</param-name>
    <param-value>
        testlib:css/bootstrap.min.css=http://cdn.bootcss.com/bootstrap/3.3.0/css/bootstrap.min.css,
        testlib:js/bootstrap.min.js=http://apps.bdimg.com/libs/bootstrap/3.3.0/js/bootstrap.min.js,
        testlib:js/jquery1.js=http://ajax.aspnetcdn.com/ajax/jQuery/testjquery-2.1.1.js
    </param-value>
</context-param>
```
```html
<head>
	<h:outputStylesheet name="css/bootstrap.min.css" library="testlib" />
	<h:outputScript name="js/jquery1.js" library="testlib" />
	<h:outputScript name="js/bootstrap.min.js" library="testlib" />
</head>
```

Have tried a raw debug, find the DefaultResourceHandler constructor is invoked but CDNResourceHandler is not (debug break point did not stop in):
```java
public CDNResourceHandler(ResourceHandler wrapped) {
		super(wrapped);
		disabledParam = getInitParameter(PARAM_NAME_CDN_DISABLED);
		cdnResources = initCDNResources();

		if (cdnResources == null) {
			throw new IllegalArgumentException(ERROR_MISSING_INIT_PARAM);
		}
	}
```

This time the console has no output info like 'Resource referenced by resourceName js/jquery1.js and libraryName testlib not found in call to ResourceHandler.createResource. It will be silenty ignored.' as before like the post above.
Do I need to change some config for the 2.1 snapshot update or can you please help to have a check? Thanks a lot!

**Comment by BalusC** [comment]
Do those resources exist in the webapp? Is the JSF project stage set to production? 

The `CDNResourceHandler` is intented to change URLs of (auto-included) JSF resources to a CDN URL. The `CDNResourceHandler` was never intented to permanently declare external resources in a JSF page. For that you can just use plain HTML `<link>` or `<script>` elements.

Sorry for any confusion caused by this. I'll clarify the `CDNResourceHandler` javadoc further.

**Comment by CatWingHub** [comment]
Hi, I see it now.
And have tried again and found 
```xml
<application>
		<resource-handler>org.omnifaces.resourcehandler.CDNResourceHandler</resource-handler>
</application>
```
is commoned in my jsf config in the recent try, after adding it, the CDNResourceHandler is invoked now.

Also have try to add an empty file in local resources, the cdn resources can be rendered in pages now, although it might not be the intention of the CDNResourceHandler.
Thanks a lot for  it!
